delivery_tnt_oca/models/delivery_carrier.py

```diff
@@ -0,0 +1,170 @@
+            else:
+                item.tnt_product_service = item.tnt_product_service_n
+
+    def tnt_oca_rate_shipment(self, order):
+        tnt_request = TntRequest(self, order)
+        response = tnt_request.rate_shipment()
+        if response["success"]:
+            response["price"] = self._tnt_oca_get_response_price(
+                response, order.currency_id, order.company_id
+            )
+        return {
+            "success": response["success"],
+            "price": response["price"],
+            "error_message": False,
+            "warning_message": False,
+        }
+
+    def _tnt_oca_get_response_price(self, response, currency, company):
+        """We need to convert the price if the currency is different."""
+        price = float(response["price"])
+        if response["currency"] != currency.name:
+            price = currency._convert(
+                price,
+                self.env["res.currency"].search([("name", "=", response["currency"])]),
+                company,
+                fields.Date.today(),
+            )
+        return price
+
+    def tnt_oca_action_label(self, picking):
+        report_name = "delivery_tnt_oca.label_delivery_tnt_oca_template"
+        iar = self.env["ir.actions.report"]
+        res = iar._get_report_from_name(report_name).render_qweb_text(picking.ids)
+        return self.env["ir.attachment"].create(
+            {
+                "name": "TNT-%s.zpl" % picking.carrier_tracking_ref,
+                "type": "binary",
+                "datas": base64.b64encode(res[0]),
+                "res_model": picking._name,
+                "res_id": picking.id,
+            }
+        )
+
+    def tnt_oca_send_shipping(self, pickings):
+        return [self.tnt_oca_create_shipping(p) for p in pickings]
+
+    def tnt_oca_create_shipping(self, picking):
+        self.ensure_one()
+        tnt_request = TntRequest(self, picking)
+        tnt_request._send_shipping()
+        tnt_request._get_label_info()
+        self.tnt_oca_action_label(picking)
+        return {
+            "exact_price": 0,
+            "tracking_number": picking.carrier_tracking_ref,
+        }
+
+    def tnt_oca_tracking_state_update(self, picking):
+        self.ensure_one()
+        if picking.carrier_tracking_ref:
+            tnt_request = TntRequest(self, picking)
+            response = tnt_request.tracking_state_update()
+            picking.delivery_state = response["delivery_state"]
+            picking.tracking_state_history = response["tracking_state_history"]
+
+    def tnt_oc_cancel_shipment(self, pickings):
+        raise NotImplementedError(
+            _("""TNT API does not allow you to cancel a shipment.""")
+        )
+
+    def tnt_oca_get_tracking_link(self, picking):
+        return "%s/%s?searchType=con&cons=%s" % (
+            "https://www.tnt.com",
+            "express/es_es/site/herramientas-envio/seguimiento.html",
+            picking.carrier_tracking_ref,
+        )
```

Comment:
About all these methods, please check if they have to be called from UI and, otherwise, prefix them with `_` to make them private.

Reply:
The `tnt_oca_tracking_state_update` method is correct according to https://github.com/OCA/delivery-carrier/blob/13.0/delivery_state/models/stock_picking.py#L64 (this is done the same in OCA carrier addons that use the dependency on `delivery_state`).

Reply:
The names of these methods are correct as defined in https://github.com/odoo/odoo/blob/13.0/addons/delivery/models/delivery_carrier.py#L24-L30